prov/gni/src/gnix_dom.c

```diff
@@ -49,13 +49,130 @@ uint32_t gnix_def_gni_rx_cq_size = 16384;
 /* TODO: should we use physical pages for gni cq rings? This is a question for Zach */
 gni_cq_mode_t gnix_def_gni_cq_modes = GNI_CQ_PHYS_PAGES;
 
+static int gnix_cm_nic_free(struct gnix_cm_nic *cm_nic)
+{
+	int ret = FI_SUCCESS;
+	gni_return_t status;
+
+	if (cm_nic == NULL)
+		return -FI_EINVAL;
+
+	atomic_dec(&cm_nic->ref_cnt);
```

Comment:
You could let reference counting handle the deallocation of the resources.
Just an example.

// take a reference
static inline gnix_cm_nic_hold(struct gnix_cm_nic *cm_nic) {
    atomic_inc(&cm_nic->ref_cnt);
}

// return a reference
static inline gnix_cm_nic_put(struct gnix_cm_nic *cm_nic) {
    if (atomic_dec_and_test(&cm_nic->ref_cnt))
	__gnix_cm_nic_release(); // Just the destructor for the cm_nic object. I
don't know the name.
}


-- Jim




On 4/3/15, 1:42 AM, "Sung-Eun Choi" <notifications@github.com> wrote:

>In 
>prov/gni/src/gnix_dom.c
><https://github.com/ofi-cray/libfabric-cray/pull/61#discussion_r27707104>:
>> @@ -49,13 +49,130 @@ uint32_t gnix_def_gni_rx_cq_size = 16384;
>>  /* TODO: should we use physical pages for gni cq rings? This is a
>>question for Zach */
>>  gni_cq_mode_t gnix_def_gni_cq_modes = GNI_CQ_PHYS_PAGES;
>>  
>> +static int gnix_cm_nic_free(struct gnix_cm_nic *cm_nic)
>> +{
>> +	int ret = FI_SUCCESS;
>> +	gni_return_t status;
>> +
>> +	if (cm_nic == NULL)
>> +		return -FI_EINVAL;
>> +
>> +	atomic_dec(&cm_nic->ref_cnt);
>Potential double free situation. Save return value from atomic_dec to
>make sure only one thread cleans up.
>—
>Reply to this email directly or
>view it on GitHub 
><https://github.com/ofi-cray/libfabric-cray/pull/61/files#r27707104>.
>
>